### DISCUSSION TITLE
Creating and updating disk images as part of gem5 resources

### DISCUSSION THREAD

**Discussion by Harshil2107** [gem5/gem5#741] [gem5-dev]
**Gem5 Disk Image Updates**

We're aiming to create new disk images as part of gem5 resources. Below, I've outlined the basic information, requirements, and desired disk images. Your input, suggestions, or alterations are most welcome.

**Requirements:**

1. **Reproducibility:** The ability to recreate the exact same disk image.
2. **Source Code Availability:** All necessary source code should reside in the gem5-resources repository, enabling the construction of the disk image without external sources.
3. **Ease of Building:** A single command or minimal set of commands should suffice.
4. **Documentation:** Comprehensive README and BUILDING markdown files with specific details:
   - README.md should encompass benchmarks on the disk image, available workloads and suites, along with annotations and compatibility details with scripts or boards.
   - BUILDING.md should guide on disk image construction.
5. **Compactness:** Disk images should strive to be as compact as possible.
6. **Realism:** Utilize stock OSes, compilers, etc., whenever feasible.
7. **Extensibility:** The ease of augmenting the base disk image with new benchmarks or alterations.
8. **Upgradability:** The ability to update disk images when newer Ubuntu versions are released.
9. **Testing:** Each benchmark should be testable, including the validation of annotations if present.

**Terminology:**

- **Benchmark:** Typically a binary or application (e.g., "cg" or "gcc").
- **Workload:** Comprehensive data necessary for benchmark execution, encompassing disk image, kernel, inputs, binaries, etc.
- **Suite:** A collection of workloads.
- **Annotation:** An instruction or mechanism executed by the benchmark/workload, conveying information to the simulator.

**Disk Image List:**

1. **Generic Ubuntu Command Line Boot:**
   - Versions for x86, Arm, and RISC-V.
   - Sequence: Arm (Kyle's version?), x86, then RISC-V.
   - Ubuntu version 22.04 with a "boot exit" workload.
   - Links:
     - [x86 Ubuntu](https://github.com/gem5/gem5-resources/tree/stable/src/x86-ubuntu)
     - [Arm Ubuntu](https://github.com/gem5/gem5-resources/tree/stable/src/arm-ubuntu)
     - [RISC-V Ubuntu](https://github.com/gem5/gem5-resources/tree/stable/src/riscv-ubuntu)

2. **NAS Parallel Benchmarks:**
   - Initially for X86, then RISC-V and Arm.
   - Built from Generic Ubuntu.
   - Link: [NAS Parallel Benchmarks](https://github.com/gem5/gem5-resources/tree/stable/src/npb)

3. **SPEC 2006 and 2017:**
   - Versions for Arm, RISC-V, and x86.
   - Built from Generic Ubuntu.
   - Includes all SPEC benchmarks, diverse workload sizes, and suite documentation.
   - Links: [SPEC 2006](https://github.com/gem5/gem5-resources/tree/stable/src/spec-2006), [SPEC 2017](https://github.com/gem5/gem5-resources/tree/stable/src/spec-2017)
   
**Note:** As we can note distribute SPEC diskimages, the goal would be to have a concise and clear readme that highlights the steps to make the diskimage for users with access to SPEC. 

4. **GAPBS, Generic HPC, Stream, Spatter:**
   - Links: [GAPBS](https://github.com/gem5/gem5-resources/tree/stable/src/gapbs), [Generic HPC](#), [Stream](#), [Spatter](#)

5. **GPU FS:**
   - Link: [GPU FS](https://github.com/gem5/gem5-resources/tree/stable/src/gpu-fs)

**Disk Image Requirements:**
- The disk should have the `m5` utility installed 

- [Nice to have] Annotation for Linux kernel finished booting, but the OS is not fully ready 
- [Required] Annotation when systemd is done and/or you can run things on the command line 
This is probably part of the base image, not the benchmark images 
- [Required] A way to programmatically run a command or commands 
- [Required] Annotation when the benchmark is starting 
- [Required] Annotation when the benchmark is done initializing 
- [Required] Annotation when the benchmark is complete 
- Better message when std is done 
- [Optional] Workload-specific annotations (e.g., for each loop iteration) 
- Script to run/test each workload 
   - [Optional] via suites 
- [Required] Remove motd (message of the day) 

Feel free to provide feedback or additional insights.

**Comment by pranith** [comment]
I am not sure this is the right place to ask, but is it possible to add simpoint markers for certain suites of benchmarks (like spec?). This will ensure that for the same versions of benchmarks, the results can be directly compared. It would also be great if we could add the scripts to create simpoints/checkpoints for the benchmarks.

One additional idea is to see if we could add result parsing scripts to generate overall score/IPC for a suite of benchmarks. For example, after running specint2k6, we can summarize the score for that run to be x/GHz or something similar.
